### PR TITLE
first part

### PR DESCRIPTION
**Pull request by EvgeniyD1**
added the main entities, controllers, services and mappers, also implemented security based on the jwt token

### CODE REVIEW

File: main_service/src/main/java/com/example/main_service/aspect/ServiceLogAspect.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public Object logAroundMethods(ProceedingJoinPoint joinPoint) throws Throwable {
+
+        StopWatch sw = new StopWatch();
+        log.info("Method " + joinPoint.getSignature().getDeclaringTypeName() + " " + joinPoint.getSignature().getName()
+                + " start");
+        sw.start();
+        Object proceed = joinPoint.proceed();
+        sw.stop();
+        log.info("Method " + joinPoint.getSignature().getDeclaringTypeName() + " " + joinPoint.getSignature().getName()
+                + " finished. Method execution time: " + sw.getTotalTimeMillis() + " milliseconds.");
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/aspect/ServiceLogAspect.java:30]: Pls use `debug` for logging. This kind is absolutely not info

File: main_service/src/main/resources/application.properties [hunk 2]
```diff
@@ -1 +1,17 @@
+spring.datasource.driver-class-name=org.postgresql.Driver
+spring.datasource.url=jdbc:postgresql://localhost:5432/main_service
+spring.datasource.username=postgres
+spring.datasource.password=root
```
Comment (skalashynski) [inline main_service/src/main/resources/application.properties:4]: Use env vars

File: main_service/src/main/resources/application.properties [hunk 3]
```diff
@@ -1 +1,17 @@
...
 
+mybatis.configuration.map-underscore-to-camel-case=true
+
+logging.level.org.springframework=WARN
+logging.level.com.example.main_service.mapper.UserMapper=TRACE
+logging.level.com.example.main_service.mapper.ArticleMapper=TRACE
+
+#24 hours
+LIFE_TIME=86400000
```
Comment (skalashynski) [inline main_service/src/main/resources/application.properties:13]: Incorrect var name `LIFE_TIME`

File: main_service/src/main/resources/application.properties [hunk 4]
```diff
@@ -1 +1,17 @@
+spring.datasource.driver-class-name=org.postgresql.Driver
+spring.datasource.url=jdbc:postgresql://localhost:5432/main_service
+spring.datasource.username=postgres
+spring.datasource.password=root
 
+mybatis.configuration.map-underscore-to-camel-case=true
+
+logging.level.org.springframework=WARN
+logging.level.com.example.main_service.mapper.UserMapper=TRACE
+logging.level.com.example.main_service.mapper.ArticleMapper=TRACE
```
Comment (skalashynski) [inline main_service/src/main/resources/application.properties:10]: Not sure about `trace` for mappers

File: main_service/src/main/java/com/example/main_service/controller/ArticleController.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    @Operation(summary = "find article by id", description = "returns article by id", responses = {
+            @ApiResponse(responseCode = "200", description = "Success!!!"),
+    })
+    @Parameter(name = "id", description = "article id to search",
+            in = ParameterIn.PATH, required = true, schema = @Schema(type = "string", defaultValue = "1"))
+    @GetMapping("{id}")
+    public ResponseEntity<Article> findArticleById(@PathVariable("id") Long id) {
+        return ResponseEntity.ok(articleService.findArticleById(id));
+    }
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/controller/ArticleController.java:48]: what if no article is found? which HTTP status should be returned?

File: main_service/src/main/java/com/example/main_service/config/JwtTokenFilter.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+        if (token != null && jwtTokenUtil.validateToken(token)) {
+            String username = jwtTokenUtil.getUsernameFromToken(token);
+            UserDetails userDetails = userService.loadUserByUsername(username);
+            Authentication authentication = new UsernamePasswordAuthenticationToken(userDetails,
+                    null, userDetails.getAuthorities());
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+        }
+
+        filterChain.doFilter(request, response);
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/config/JwtTokenFilter.java:40]: If the is no token, why do you go later? why do you call `doFilter`?

File: main_service/src/main/java/com/example/main_service/controller/ArticleController.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    @Operation(summary = "find article by title with author", description = "returns article by title with author like",
+            responses = {
+            @ApiResponse(responseCode = "200", description = "Success!!!"),
+    })
+    @Parameter(name = "title", description = "article title to search",
+            in = ParameterIn.PATH, required = true, schema = @Schema(type = "string", defaultValue = "text"))
+    @GetMapping("/byTitleWithAuthor/{title}")
+    public ResponseEntity<List<Article>> findArticleByTitleWithAuthor(@PathVariable("title") String title){
+        return ResponseEntity.ok(articleService.findArticlesByTitleWithAuthor(title));
+    }
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/controller/ArticleController.java:168]: what if no article is found? which HTTP status should be returned?

File: main_service/src/main/java/com/example/main_service/controller/RegistrationController.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Operation(summary = "registration form", description = "write your details here", responses = {
+            @ApiResponse(responseCode = "200", description = "Success!!!"),
+    })
+    @PostMapping("/registration")
+    public ResponseEntity<String> createAccount(@RequestBody SignRequest request){
+        if (userService.findUserByUsername(request.getUsername())!=null){
+            return ResponseEntity.badRequest().body("A user with the same name already exists!");
+        }
+        userService.insertUser(request);
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/controller/RegistrationController.java:36]: you have to save all username with the lower_case and validate respectfully.

File: main_service/src/main/java/com/example/main_service/mapper/ArticleMapper.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.apache.ibatis.mapping.FetchType;
+
+import java.util.List;
+
+@Mapper
+public interface ArticleMapper {
+
+    @Select("select * from article where id = #{id}")
+    Article findArticleById(@Param("id") Long id);
+
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/mapper/ArticleMapper.java:23]: Is it possible to return `Optional<Article>`?

File: main_service/src/main/java/com/example/main_service/service/ArticleService.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    public List<Article> findArticlesByUserId(Long userId){
+        return articleMapper.findArticlesByUserId(userId);
+    }
+
+    public void insertArticle(Long userId, ArticleRequest request){
+        Article article = new Article();
+        article.setUserId(userId);
+        article.setTitle(request.getTitle());
+        article.setContext(request.getContext());
+        article.setCreationDate(new Date());
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/service/ArticleService.java:36]: setCreationDate -> setCreationDateTime

File: main_service/src/main/java/com/example/main_service/service/ArticleService.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    public List<Article> findArticlesByTitle(String title){
+        return articleMapper.findArticlesByTitle(title);
+    }
+
+    public List<Article> findArticlesByUserId(Long userId){
+        return articleMapper.findArticlesByUserId(userId);
+    }
+
+    public void insertArticle(Long userId, ArticleRequest request){
+        Article article = new Article();
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/service/ArticleService.java:32]: Let me know, how do you open transaction?

File: main_service/src/main/java/com/example/main_service/service/UserService.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        int userCount = userMapper.getTotalUserCount();
+        int totalPages = (userCount + size - 1) / size;
+        return new UserPageRequest(users, totalPages, page, size);
+    }
+
+    public void insertUser(SignRequest request) {
+        User user = new User();
+        user.setUsername(request.getUsername());
+        user.setPassword(passwordEncoder.encode(request.getPassword()));
+        user.setRegistrationDate(new Date());
```
Comment (skalashynski) [inline main_service/src/main/java/com/example/main_service/service/UserService.java:52]: use datetime

**Review by skalashynski — CHANGES_REQUESTED** [review]

File: main_service/src/main/java/com/example/main_service/aspect/ServiceLogAspect.java (hunk 1 above)
Comment (EvgeniyD1) [reply main_service/src/main/java/com/example/main_service/aspect/ServiceLogAspect.java:30]: what I should do?

File: main_service/src/main/java/com/example/main_service/domain/Article.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.Date;
+
+@Data
+public class Article implements Serializable {
+
+    private Long id;
+    private Long userId;
+    private String title;
+    private String context;
+    private Date creationDate;
```
[the remark this answers is not in the document]
Comment (EvgeniyD1) [reply main_service/src/main/java/com/example/main_service/domain/Article.java:15]: fixed

File: main_service/src/main/resources/application.properties [hunk 14]
```diff
@@ -1 +1,13 @@
...
+
+#24 hours
+LIFE_TIME=86400000
```
[the remark this answers is not in the document]
Comment (EvgeniyD1) [reply main_service/src/main/resources/application.properties]: it's jwt token lifetime

File: main_service/src/main/java/com/example/main_service/MainServiceApplication.java [hunk 15]
```diff
@@ -2,8 +2,10 @@
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
 
 @SpringBootApplication
+@EnableTransactionManagement(proxyTargetClass = true)
```
[the remark this answers is not in the document]
Comment (EvgeniyD1) [reply main_service/src/main/java/com/example/main_service/MainServiceApplication.java:8]: in's need to work  @PreAuthorize()

File: main_service/src/main/resources/application.properties (hunk 2 above)
Comment (EvgeniyD1) [reply main_service/src/main/resources/application.properties:4]: fixed

File: main_service/src/main/resources/application.properties (hunk 3 above)
Comment (EvgeniyD1) [reply main_service/src/main/resources/application.properties:13]: what does that mean in this case?

File: main_service/src/main/resources/application.properties (hunk 4 above)
Comment (EvgeniyD1) [reply main_service/src/main/resources/application.properties:10]: ok

File: main_service/src/main/java/com/example/main_service/config/JwtTokenFilter.java (hunk 6 above)
Comment (EvgeniyD1) [reply main_service/src/main/java/com/example/main_service/config/JwtTokenFilter.java:40]: ok, what i should to do?

**Review by skalashynski — APPROVED** [review]

**Pull request merged**